contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,327 @@
+    mapping(IERC20Upgradeable => uint256) public lastRewardBalance;
+
+    address public feeCollector;
+
+    /// @notice The deposit fee, scaled to `PRECISION`
+    uint256 public depositFeePercent;
+
+    /// @notice Accumulated `token` rewards per share, scaled to `PRECISION`. See above
+    mapping(IERC20Upgradeable => uint256) public accRewardPerShare;
+    /// @notice `PRECISION` of `accRewardPerShare`
+    uint256 public PRECISION;
+
+    /// @dev Info of each user that stakes JOE
+    mapping(address => UserInfo) private userInfo;
+
+    /// @notice Emitted when a user deposits JOE
+    event Deposit(address indexed user, uint256 amount, uint256 fee);
+
+    /// @notice Emitted when owner changes the deposit fee percentage
+    event DepositFeeChanged(uint256 newFee, uint256 oldFee);
+
+    /// @notice Emitted when a user withdraws JOE
+    event Withdraw(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user emergency withdraws its JOE
+    event EmergencyWithdraw(address indexed user, uint256 amount);
+
+    /// @notice Emitted when owner adds a token to the reward tokens list
+    event RewardTokenAdded(address token);
+
+    /// @notice Emitted when owner removes a token to the reward tokens list
+    event RewardTokenRemoved(address token);
+
+    /**
+     * @notice Initialize a new StableJoeStaking contract
+     * @dev This contract needs to receive an ERC20 `_rewardToken` in order to distribute them
+     * (with JoeMakerV4 in our case)
+     * @param _rewardToken The address of the ERC20 reward token
+     * @param _joe The address of the JOE token
+     * @param _depositFeePercent The deposit fee percent, scalled to 1e18, e.g. 3% is 3e16
+     */
+    function initialize(
+        IERC20Upgradeable _rewardToken,
+        IERC20Upgradeable _joe,
+        address _feeCollector,
+        uint256 _depositFeePercent
+    ) external initializer {
+        __Ownable_init();
+        require(_feeCollector != address(0), "StableJoeStaking: fee collector can't be address 0");
+        require(_depositFeePercent < 5e17, "StableJoeStaking: max deposit fee can't be greater than 50%");
+
+        joe = _joe;
+        depositFeePercent = _depositFeePercent;
+        feeCollector = _feeCollector;
+
+        isRewardToken[_rewardToken] = true;
+        rewardTokens.push(_rewardToken);
+        PRECISION = 1e24;
+    }
+
+    /**
+     * @notice Deposit JOE for reward token allocation
+     * @param _amount The amount of JOE to deposit
+     */
+    function deposit(uint256 _amount) external {
+        UserInfo storage user = userInfo[msg.sender];
+
+        uint256 _fee = (_amount * depositFeePercent) / 1e18;
+        uint256 _amountMinusFee = _amount.sub(_fee);
+
+        uint256 _previousAmount = user.amount;
+        uint256 _newAmount = user.amount.add(_amountMinusFee);
+        user.amount = _newAmount;
+
+        uint256 _len = rewardTokens.length;
+        uint256 _pending;
```

Comment:
Why does `_pending` need to be initialised outside of for loop? I assume to prevent stack too deep?

Reply:
Not at all this doesn't prevent stack too deep error in any way, but I always learned that it's better to do it that way for memory generally speaking.
Tho, I don't know if it's useful in solidity and if it uses less gas